From f6c70adf76aba8ccfc1e6b26de3e47ae85af1cee Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Wed, 15 Oct 2025 13:51:11 +0200
Subject: [PATCH] wifi: brcmfmac: sdio: fix reference counting use-after-free
 on reboot

On system reboot brcmf_sdio_remove triggers a use-after-free by calling
kthread_stop [1].

This is likely because kthread_run does not increment reference counter
on the task handle (not verified).

For now this issue can be resolved by calling get_task_struct after
kthread_run to increment the reference counter once.

[1]
[  188.764642] refcount_t: addition on 0; use-after-free.
[  188.769820] WARNING: CPU: 3 PID: 536 at lib/refcount.c:25 refcount_warn_saturate+0xc8/0x144
[  188.778187] Modules linked in: xt_conntrack xt_MASQUERADE xfrm_user xfrm_algo iptable_nat nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 xt_addrtype iptable_filter ip_tables x_tables br_netfilter overlay algif_hash algif_skcipher af_alg brcmfmac_cyw fsl_jr_uio caam_jr caamkeyblob_desc brcmfmac caamhash_desc caamalg_desc crypto_engine authenc libdes hci_uart cfg80211 crct10dif_ce btqca polyval_ce snd_soc_fsl_asoc_card onboard_usb_hub btbcm polyval_generic snd_soc_imx_audmux snd_soc_imx_hdmi brcmutil bluetooth dw_hdmi_cec snd_soc_fsl_sai snd_soc_fsl_aud2htx snd_soc_fsl_utils imx_pcm_dma snd_soc_wm8904 ecdh_generic imx_sdma rtc_abx80x at24 ecc caam rfkill_gpio secvio error rfkill
[  188.838610] CPU: 3 PID: 536 Comm: init Not tainted 6.6.52-06828-g52bedfb5495d #47
[  188.846096] Hardware name: SolidRun i.MX8MP HummingBoard Pulse (DT)
[  188.852363] pstate: 60000005 (nZCv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[  188.859328] pc : refcount_warn_saturate+0xc8/0x144
[  188.864126] lr : refcount_warn_saturate+0xc8/0x144
[  188.868921] sp : ffff800082dcba40
[  188.872238] x29: ffff800082dcba40 x28: ffff0000c670d880 x27: 0000000000000000
[  188.879381] x26: 0000000000000000 x25: 0000000000000000 x24: ffff800081f3e780
[  188.886526] x23: ffff0000c03af888 x22: ffff0000c0b49488 x21: ffff0000c6179800
[  188.893669] x20: ffff0000c19e49e8 x19: ffff0000c19e49c0 x18: 0000000000000006
[  188.900816] x17: 0000000000000000 x16: 0000000000000000 x15: ffff800082dcb4c0
[  188.907958] x14: 0000000000000000 x13: ffff800081f41798 x12: 0000000000000651
[  188.915103] x11: 000000000000021b x10: ffff800081f99798 x9 : ffff800081f41798
[  188.922246] x8 : 00000000ffffefff x7 : ffff800081f99798 x6 : 80000000fffff000
[  188.929391] x5 : 000000000000bff4 x4 : 0000000000000000 x3 : 0000000000000000
[  188.936535] x2 : 0000000000000000 x1 : 0000000000000000 x0 : ffff0000c670d880
[  188.943682] Call trace:
[  188.946129]  refcount_warn_saturate+0xc8/0x144
[  188.950579]  kthread_stop+0xd8/0x198
[  188.954157]  brcmf_sdio_remove+0x48/0x408 [brcmfmac]
[  188.959190]  brcmf_sdiod_remove+0x24/0xa8 [brcmfmac]
[  188.964212]  brcmf_ops_sdio_remove+0x54/0x94 [brcmfmac]
[  188.969493]  sdio_bus_remove+0x3c/0x16c
[  188.973336]  device_remove+0x4c/0x80
[  188.976920]  device_release_driver_internal+0x1d4/0x230
[  188.982149]  driver_detach+0x50/0x98
[  188.985729]  bus_remove_driver+0x6c/0xbc
[  188.989658]  driver_unregister+0x30/0x60
[  188.993583]  sdio_unregister_driver+0x24/0x30
[  188.997944]  brcmf_sdio_exit+0x18/0x24 [brcmfmac]
[  189.002701]  brcmf_core_exit+0x10/0x1c [brcmfmac]
[  189.007460]  brcmf_reboot_callback+0x20/0x30 [brcmfmac]
[  189.012741]  blocking_notifier_call_chain+0x6c/0xa0
[  189.017624]  kernel_restart+0x28/0xc0
[  189.021295]  __do_sys_reboot+0x154/0x25c
[  189.025224]  __arm64_sys_reboot+0x24/0x30
[  189.029241]  invoke_syscall+0x48/0x118
[  189.032995]  el0_svc_common.constprop.0+0x40/0xe0
[  189.037705]  do_el0_svc+0x1c/0x28
[  189.041025]  el0_svc+0x40/0xe8
[  189.044086]  el0t_64_sync_handler+0x100/0x12c
[  189.048449]  el0t_64_sync+0x190/0x194
[  189.052117] ---[ end trace 0000000000000000 ]---
[  189.056788] Unable to handle kernel NULL pointer dereference at virtual address 0000000000000000
[  189.065590] Mem abort info:
[  189.068384]   ESR = 0x0000000096000004
[  189.072141]   EC = 0x25: DABT (current EL), IL = 32 bits
[  189.077459]   SET = 0, FnV = 0
[  189.080512]   EA = 0, S1PTW = 0
[  189.083660]   FSC = 0x04: level 0 translation fault
[  189.088548] Data abort info:
[  189.091436]   ISV = 0, ISS = 0x00000004, ISS2 = 0x00000000
[  189.096926]   CM = 0, WnR = 0, TnD = 0, TagAccess = 0
[  189.101986]   GCS = 0, Overlay = 0, DirtyBit = 0, Xs = 0
[  189.107304] user pgtable: 4k pages, 48-bit VAs, pgdp=00000001018b4000
[  189.113751] [0000000000000000] pgd=0000000000000000, p4d=0000000000000000
[  189.120552] Internal error: Oops: 0000000096000004 [#1] PREEMPT SMP
[  189.126822] Modules linked in: xt_conntrack xt_MASQUERADE xfrm_user xfrm_algo iptable_nat nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 xt_addrtype iptable_filter ip_tables x_tables br_netfilter overlay algif_hash algif_skcipher af_alg brcmfmac_cyw fsl_jr_uio caam_jr caamkeyblob_desc brcmfmac caamhash_desc caamalg_desc crypto_engine authenc libdes hci_uart cfg80211 crct10dif_ce btqca polyval_ce snd_soc_fsl_asoc_card onboard_usb_hub btbcm polyval_generic snd_soc_imx_audmux snd_soc_imx_hdmi brcmutil bluetooth dw_hdmi_cec snd_soc_fsl_sai snd_soc_fsl_aud2htx snd_soc_fsl_utils imx_pcm_dma snd_soc_wm8904 ecdh_generic imx_sdma rtc_abx80x at24 ecc caam rfkill_gpio secvio error rfkill
[  189.187197] CPU: 3 PID: 536 Comm: init Tainted: G        W          6.6.52-06828-g52bedfb5495d #47
[  189.196160] Hardware name: SolidRun i.MX8MP HummingBoard Pulse (DT)
[  189.202427] pstate: 60000005 (nZCv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[  189.209394] pc : kthread_stop+0x150/0x198
[  189.213409] lr : kthread_stop+0xd8/0x198
[  189.217336] sp : ffff800082dcba50
[  189.220651] x29: ffff800082dcba50 x28: ffff0000c670d880 x27: 0000000000000000
[  189.227797] x26: 0000000000000000 x25: 0000000000000000 x24: ffff800081f3e780
[  189.234943] x23: ffff0000c03af888 x22: ffff0000c0b49488 x21: 0000000000000000
[  189.242085] x20: ffff0000c19e49e8 x19: ffff0000c19e49c0 x18: 0000000000000006
[  189.249228] x17: 0000000000000000 x16: 0000000000000000 x15: ffff800082dcb4c0
[  189.256372] x14: 0000000000000000 x13: ffff800081f41798 x12: 0000000000000651
[  189.263517] x11: 000000000000021b x10: ffff800081f99798 x9 : ffff800081f41798
[  189.270661] x8 : 00000000ffffefff x7 : ffff800081f99798 x6 : 80000000fffff000
[  189.277804] x5 : 000000000000bff4 x4 : 0000000000000000 x3 : 0000000000000000
[  189.284948] x2 : 0000000000000000 x1 : 0000000000000000 x0 : 000000000020804c
[  189.292095] Call trace:
[  189.294541]  kthread_stop+0x150/0x198
[  189.298205]  brcmf_sdio_remove+0x48/0x408 [brcmfmac]
[  189.303225]  brcmf_sdiod_remove+0x24/0xa8 [brcmfmac]
[  189.308244]  brcmf_ops_sdio_remove+0x54/0x94 [brcmfmac]
[  189.313525]  sdio_bus_remove+0x3c/0x16c
[  189.317365]  device_remove+0x4c/0x80
[  189.320948]  device_release_driver_internal+0x1d4/0x230
[  189.326180]  driver_detach+0x50/0x98
[  189.329758]  bus_remove_driver+0x6c/0xbc
[  189.333688]  driver_unregister+0x30/0x60
[  189.337613]  sdio_unregister_driver+0x24/0x30
[  189.341973]  brcmf_sdio_exit+0x18/0x24 [brcmfmac]
[  189.346732]  brcmf_core_exit+0x10/0x1c [brcmfmac]
[  189.351490]  brcmf_reboot_callback+0x20/0x30 [brcmfmac]
[  189.356770]  blocking_notifier_call_chain+0x6c/0xa0
[  189.361654]  kernel_restart+0x28/0xc0
[  189.365322]  __do_sys_reboot+0x154/0x25c
[  189.369252]  __arm64_sys_reboot+0x24/0x30
[  189.373268]  invoke_syscall+0x48/0x118
[  189.377022]  el0_svc_common.constprop.0+0x40/0xe0
[  189.381733]  do_el0_svc+0x1c/0x28
[  189.385053]  el0_svc+0x40/0xe8
[  189.388112]  el0t_64_sync_handler+0x100/0x12c
[  189.392477]  el0t_64_sync+0x190/0x194
[  189.396146] Code: c8017e60 35ffffa1 17ffffc7 f98002b1 (c85f7ea0)

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
index db9f45e..870e5c4 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
@@ -2884,12 +2884,14 @@ static void brcmf_sdio_bus_stop(struct device *dev)
 	if (bus->watchdog_tsk) {
 		send_sig(SIGTERM, bus->watchdog_tsk, 1);
 		kthread_stop(bus->watchdog_tsk);
+		put_task_struct(bus->watchdog_tsk);
 		bus->watchdog_tsk = NULL;
 	}
 
 	if (bus->thr_rxf_ctl.p_task) {
 		send_sig(SIGTERM, bus->thr_rxf_ctl.p_task, 1);
 		kthread_stop(bus->thr_rxf_ctl.p_task);
+		put_task_struct(bus->thr_rxf_ctl.p_task);
 		bus->thr_rxf_ctl.p_task = NULL;
 	}
 
@@ -5715,6 +5717,8 @@ struct brcmf_sdio *brcmf_sdio_probe(struct brcmf_sdio_dev *sdiodev)
 		if (IS_ERR(bus->thr_rxf_ctl.p_task)) {
 			brcmf_err("brcmf_sdio_rxf_thread failed to start\n");
 			bus->thr_rxf_ctl.p_task = NULL;
+		} else {
+			get_task_struct(bus->thr_rxf_ctl.p_task);
 		}
 	}
 
@@ -5728,6 +5732,8 @@ struct brcmf_sdio *brcmf_sdio_probe(struct brcmf_sdio_dev *sdiodev)
 	if (IS_ERR(bus->watchdog_tsk)) {
 		pr_warn("brcmf_watchdog thread failed to start\n");
 		bus->watchdog_tsk = NULL;
+	} else {
+		get_task_struct(bus->watchdog_tsk);
 	}
 	/* Initialize DPC thread */
 	bus->dpc_triggered = false;
@@ -5799,12 +5805,14 @@ void brcmf_sdio_remove(struct brcmf_sdio *bus)
 		if (bus->watchdog_tsk) {
 			send_sig(SIGTERM, bus->watchdog_tsk, 1);
 			kthread_stop(bus->watchdog_tsk);
+			put_task_struct(bus->watchdog_tsk);
 			bus->watchdog_tsk = NULL;
 		}
 
 		if (bus->thr_rxf_ctl.p_task) {
 			send_sig(SIGTERM, bus->thr_rxf_ctl.p_task, 1);
 			kthread_stop(bus->thr_rxf_ctl.p_task);
+			put_task_struct(bus->thr_rxf_ctl.p_task);
 			bus->thr_rxf_ctl.p_task = NULL;
 		}
 
-- 
2.51.0

